File: applications/rtkprojectgeometricphantom/rtkprojectgeometricphantom.py
import argparse
import sys
import itk
from itk import RTK as rtk
import numpy as np


def build_parser():
    # Argument parsing
    parser = argparse.ArgumentParser(
        description="Computes projections through a 3D phantom described by a file, according to a geometry"
    )
    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output projections file name", type=str, required=True
    )
    parser.add_argument(
        "--phantomfile",
        help="Configuration parameters for the phantom",
        type=str,
        required=True,
    )
    parser.add_argument(
        "--phantomscale",
        help="Scaling factor for the phantom dimensions",
        type=rtk.comma_separated_args(float),
        default=[1.0],
    )
    parser.add_argument(
        "--offset",
        help="3D spatial offset of the phantom center",
        type=rtk.comma_separated_args(float),
    )
    parser.add_argument(
        "--rotation",
        help="Rotation matrix for the phantom",
        type=rtk.comma_separated_args(float),
    )

    rtk.add_rtk3Doutputimage_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):

    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Create a stack of empty projection images
    constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)

    # Adjust size according to geometry
    sizeOutput = constantImageSource.GetSize()
    sizeOutput[2] = len(geometry.GetGantryAngles())
    constantImageSource.SetSize(sizeOutput)

    # Create SheppLogan Phantom
    offset = [0] * Dimension
    if args_info.offset is not None:
        if len(args_info.offset) > 3:
            print("--offset needs up to 3 values")
            sys.exit(1)
        for i in range(len(args_info.offset)):
            offset[i] = args_info.offset[i]

    scale = [args_info.phantomscale[0]] * 3
    if len(args_info.phantomscale) > 3:
        print("--phantomscale needs up to 3 values")
        sys.exit(1)
    for i in range(len(args_info.phantomscale)):
        scale[i] = args_info.phantomscale[i]

    rot = itk.Matrix[itk.D, Dimension, Dimension]()
    rot.SetIdentity()

    if args_info.rotation is not None:
        if len(args_info.rotation) != 9:
            print("--rotation needs exactly 9 values")
            sys.exit(1)
        itk.matrix_from_array(np.array(args_info.rotation).reshape(3, 3))

    ppc = rtk.ProjectGeometricPhantomImageFilter[OutputImageType, OutputImageType].New()
    ppc.SetInput(constantImageSource.GetOutput())
    ppc.SetGeometry(geometry)
    ppc.SetPhantomScale(scale)
    ppc.SetOriginOffset(offset)
    ppc.SetRotationMatrix(rot)
    ppc.SetConfigFile(args_info.phantomfile)
    ppc.Update()

    # Write
    if args_info.verbose:
        print("Projecting and writing... ")

    itk.imwrite(ppc.GetOutput(), args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()

File: applications/rtkprojectgeometricphantom/rtkprojectgeometricphantom.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkprojectgeometricphantom_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkRayEllipsoidIntersectionImageFilter.h"
#include "rtkProjectGeometricPhantomImageFilter.h"

#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkprojectgeometricphantom, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

  using OutputImageType = itk::Image<OutputPixelType, Dimension>;

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;

  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Create a stack of empty projection images
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkprojectgeometricphantom>(constantImageSource,
                                                                                                    args_info);

  // Adjust size according to geometry
  constantImageSource->SetSize(itk::MakeSize(
    constantImageSource->GetSize()[0], constantImageSource->GetSize()[1], geometry->GetGantryAngles().size()));

  using PPCType = rtk::ProjectGeometricPhantomImageFilter<OutputImageType, OutputImageType>;

  // Offset, scale, rotation
  PPCType::VectorType offset(0.);
  if (args_info.offset_given)
  {
    if (args_info.offset_given > 3)
    {
      std::cerr << "--offset needs up to 3 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    offset[0] = args_info.offset_arg[0];
    offset[1] = args_info.offset_arg[1];
    offset[2] = args_info.offset_arg[2];
  }
  PPCType::VectorType scale;
  scale.Fill(args_info.phantomscale_arg[0]);
  if (args_info.phantomscale_given)
  {
    if (args_info.phantomscale_given > 3)
    {
      std::cerr << "--phantomscale needs up to 3 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    for (unsigned int i = 0; i < std::min(args_info.phantomscale_given, Dimension); i++)
      scale[i] = args_info.phantomscale_arg[i];
  }
  PPCType::RotationMatrixType rot;
  rot.SetIdentity();
  if (args_info.rotation_given)
  {
    if (args_info.rotation_given != 9)
    {
      std::cerr << "--rotation needs exactly 9 values" << std::endl;
      exit(EXIT_FAILURE);
    }
    for (unsigned int i = 0; i < Dimension; i++)
      for (unsigned int j = 0; j < Dimension; j++)
        rot[i][j] = args_info.rotation_arg[i * Dimension + j];
  }

  auto ppc = PPCType::New();
  ppc->SetInput(constantImageSource->GetOutput());
  ppc->SetGeometry(geometry);
  ppc->SetPhantomScale(scale);
  ppc->SetOriginOffset(offset);
  ppc->SetRotationMatrix(rot);
  ppc->SetConfigFile(args_info.phantomfile_arg);

  TRY_AND_EXIT_ON_ITK_EXCEPTION(ppc->Update())

  // Write
  if (args_info.verbose_flag)
    std::cout << "Projecting and writing... " << std::flush;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(ppc->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}
